language/move-prover/bytecode-to-boogie/src/spec_translator.rs

```diff
@@ -0,0 +1,317 @@
+            BinOp::Mul => format!("Integer(i#Integer({}) * i#Integer({}))", left, right),
+            BinOp::Mod => format!("Integer(i#Integer({}) mod i#Integer({}))", left, right),
+            BinOp::Div => format!("Integer(i#Integer({}) div i#Integer({}))", left, right),
+            BinOp::BitOr => format!("Integer(i#Integer({}) | i#Integer({}))", left, right),
+            BinOp::BitAnd => format!("Integer(i#Integer({}) & i#Integer({}))", left, right),
+            BinOp::Xor => format!("Integer(i#Integer({}) ^ i#Integer({}))", left, right),
+
+            // bool
+            BinOp::And => format!("Boolean(b#Boolean({}) ^ b#Boolean({}))", left, right),
+            BinOp::Or => format!("Boolean(b#Boolean({}) ^ b#Boolean({}))", left, right),
+
+            // generic equality
+            BinOp::Eq => format!("Boolean(({}) == ({}))", left, right),
+            BinOp::Neq => format!("Boolean(({}) != ({}))", left, right),
+
+            // Ordering
+            // TODO: is this defined also for non-integer types?
+            BinOp::Lt => format!("Boolean(i#Integer({}) < i#Integer({}))", left, right),
+            BinOp::Gt => format!("Boolean(i#Integer({}) > i#Integer({}))", left, right),
+            BinOp::Le => format!("Boolean(i#Integer({}) <= i#Integer({}))", left, right),
+            BinOp::Ge => format!("Boolean(i#Integer({}) >= i#Integer({}))", left, right),
+        }
+    }
+
+    /// Translates a constant.
+    fn translate_constant(&mut self, val: &CopyableVal) -> String {
+        match val {
+            CopyableVal::Address(addr) => {
+                format!("Address({})", self.translate_account_address(addr))
+            }
+            CopyableVal::U64(val) => format!("Integer({})", val),
+            CopyableVal::Bool(val) => format!("Boolean({})", val),
+            // TODO: byte arrays
+            CopyableVal::ByteArray(_arr) => {
+                self.error("ByteArray not implemented", "<bytearray>".to_string())
+            }
+        }
+    }
+
+    /// Translates a location into a boogie expression of type Value.
+    ///
+    /// The StorageLocation AST can represent different types in the Boogie model,
+    /// therefore we need to interpret in context. See also `translate_location_as_reference`
+    /// and `translate_location_as_address`.
+    fn translate_location_as_value(&mut self, loc: &StorageLocation) -> String {
+        match loc {
+            StorageLocation::Formal(name) => {
+                // TODO: we like to preserve original argument names in the boogie code,
+                //   which can be achieved with a little more effort.
+                return format!("arg{}", self.get_param_index(name));
+            }
+            StorageLocation::Old(loc) => format!("old({})", self.translate_location_as_value(loc)),
+            StorageLocation::Ret => {
+                // TODO: the parser currently only handles `return` for single value. Need to
```

Comment:
Ah, good catch! Should be an easy fix in the parser if we mimic this `ret_i` approach.